src/renderer/gdi/state.cpp

```diff
@@ -406,29 +397,30 @@ GdiEngine::~GdiEngine()
         _lineMetrics.underlineOffset2 = _lineMetrics.underlineOffset - _lineMetrics.gridlineWidth;
     }
 
-    // Curly line doesn't render properly below 1px stroke width. Make it a straight line.
-    if (_lineMetrics.underlineWidth < 1)
-    {
-        _lineMetrics.curlylinePeakHeight = 0;
-    }
-    else
+    const int strokeHalfWidth = std::lround(_lineMetrics.underlineWidth / 2.0);
+
+    // Since we use GDI pen for drawing, the underline offset should point to
+    // the center of the underline.
+    _lineMetrics.underlineOffset += strokeHalfWidth;
+    _lineMetrics.underlineOffset2 += strokeHalfWidth;
+
+    // We want the underline to always be visible and remain within the cell
+    // bottom, so we clamp the offset to fit just inside.
+    _lineMetrics.underlineOffset = std::min(_lineMetrics.underlineOffset, maxUnderlineOffset);
+    _lineMetrics.underlineOffset2 = std::min(_lineMetrics.underlineOffset2, maxUnderlineOffset);

```

Comment:
This still seems wrong to me. The offsets have already been clamped prior to this, so no additional clamping should have been necessary. The `strokeHalfWidth` adjustment should just be accounting for the difference between a `PatBlt` rendering and a `LineTo` rendering. If it was correct, then the `LineTo` rendering should exactly match where the `PatBlt` would have rendered without that adjustment. If it ends up out of bounds, then that's clearly not the case, so something must be wrong with the `strokeHalfWidth` calculation. Maybe a rounding issue? I don't know.

Worst case, if you can't get it to work, at least move the `strokeHalfWidth` adjustment higher up, prior to where we're doing the first clamping (and get rid of this second set of clamping). Because we've got special case handling for the double underline offset, which is potentially going to be broken if you clamp it a second time.

Reply:
I think another potential problem is that the `maxUnderlineOffset` that we're using for clamping is intended to be used with the original `PatBlt` offsets. Once you add the `strokeHalfWidth` adjustments, that range is no longer applicable. The only way I can see this working is if we can get that `strokeHalfWidth` calculation correct.

I'd suggest writing a little test app that renders a bunch of lines of varying widths, using both `LineTo` and `PatBlt`. From that you should be able to figure out exactly what offset needs to be applied at each width to get the two renderings to match.

Reply:
Unfortunately, I'm not well-versed in Windows programming in general, so creating an app from scratch would be a different task for me 😅

However, I did experiment with the idea of comparing `LineTo` and `PatBlt`, and it made it much more clear what I needed to do.

With this code:
```cpp
// a fixed offset
const auto cellMidY = std::lround(fontHeight / 2.0f);

// adjust for drawing with DrawStrokedLine()
const auto cellMidYWithHalfStrokeWidth = cellMidY + gsl::narrow_cast<int>(std::floor(_lineMetrics.underlineWidth / 2));

// Select a brush

// draw a line with PatBlt()
RETURN_HR_IF(E_FAIL, !DrawLine(ptTarget.x, ptTarget.y + cellMidY, widthOfAllCells, _lineMetrics.underlineWidth));

// Select the dotted pen

// draw a line with LineTo()
RETURN_HR_IF(E_FAIL, !DrawStrokedLine(ptTarget.x, ptTarget.y + cellMidYWithHalfStrokeWidth, widthOfAllCells));
```

Result:

```pwsh
"`e[4:4m Dotted `e[m"
```


![Screenshot 2023-12-10 113355](https://github.com/microsoft/terminal/assets/55626797/ac1afcdf-514f-4762-80d5-51666cebe450)

![Screenshot 2023-12-10 112020](https://github.com/microsoft/terminal/assets/55626797/2f9d6e43-09cf-4eed-86f8-75960e0d82a1)

I tried this with other font sizes, and they all show the same result. So, the `DrawLine()` and `DrawStrokedLine()` calls with those args are synonymous to each other.

> This still seems wrong to me. The offsets have already been clamped prior to this, so no additional clamping should have been necessary. 

Yeah, you're right! It wouldn't have been necessary if I had floored `strokeHalfWidth` before adding it to the two offsets. (I had `lround`'d it, so that was the issue.)